### PR TITLE
[SPARK-45262][SQL][TESTS][DOCS] Improve examples for regexp parameters

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to add a few more examples for `LIKE`, `ILIKE`, `RLIKE`, `regexp_instr()`, `regexp_extract_all()` that highlight correctness of current description and test a couple more of corner cases.

### Why are the changes needed?
The description of `LIKE` says:
```
... in order to match "\abc", the pattern should be "\\abc"
```
but in Spark SQL shell:
```sql
spark-sql (default)> SELECT c FROM t;
\abc
spark-sql (default)> SELECT c LIKE "\\abc" FROM t;
[INVALID_FORMAT.ESC_IN_THE_MIDDLE] The format is invalid: '\\abc'. The escape character is not allowed to precede 'a'.
spark-sql (default)> SELECT c LIKE "\\\\abc" FROM t;
true
```
So, the description might confuse users since the pattern must contain 4 slashes when the pattern is a regular SQL string.

New example shows that the pattern "\\abc" is correct if we take into account the string as a raw string:
```sql
spark-sql (default)> SELECT c LIKE R"\\abc" FROM t;
true
```

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
By running new and modified tests:
```
$ build/sbt "test:testOnly *.StringFunctionsSuite"
$ build/sbt "sql/test:testOnly org.apache.spark.sql.expressions.ExpressionInfoSuite"
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 1]
```diff
@@ -105,13 +105,15 @@ abstract class StringRegexExpression extends BinaryExpression
...
       true
+      > SELECT '\\abc' AS S, S _FUNC_ R'\\abc', S _FUNC_ '\\\\abc';
+      \abc	true	true
       > SET spark.sql.parser.escapedStringLiterals=true;
       spark.sql.parser.escapedStringLiterals	true
       > SELECT '%SystemDrive%\Users\John' _FUNC_ '\%SystemDrive\%\\Users%';
       true
       > SET spark.sql.parser.escapedStringLiterals=false;
       spark.sql.parser.escapedStringLiterals	false
-      > SELECT '%SystemDrive%\\Users\\John' _FUNC_ '\%SystemDrive\%\\\\Users%';
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:114]: This is the example you want to fix, right?

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:114]: Do we need to change this part?

https://github.com/apache/spark/blob/fdedec1de54fa7619e1d1b59eb6f15a49e727428/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L95-L96

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:114]: it looks to me that we follow the java string semantic: if you write `"\\abc"`, you actually get `\abc`. Is it also true in simple `SELECT` not `LIKE`?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:114]: Do we have a doc to explain what the string value we get eventually by writing string literals in SQL?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:114]: I'm seeing some weird behaviors:
```
scala> sql(""" select "a\b" """).show
+---+
|a\b|
+---+
|a\b|
+---+

scala> sql(""" select "a\\b" """).show
+---+
|a\b|
+---+
|a\b|
+---+
```

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:114]: > https://github.com/apache/spark/blob/fdedec1de54fa7619e1d1b59eb6f15a49e727428/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala#L95-L96

I guess the "\abc" means the fact value but the literal. Of course, it make users confused. I agree to update the doc here to give more explanation.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:114]: > Do we have a doc to explain what the string value we get eventually by writing string literals in SQL?

There is the doc about string literals: https://spark.apache.org/docs/latest/sql-ref-literals.html#string-literal. It explain something implemented in https://github.com/apache/spark/blob/5f643ee1444d41ca78ef397d9675efac550a6967/sql/api/src/main/scala/org/apache/spark/sql/catalyst/util/SparkParserUtils.scala#L34-L54

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:114]: > This is the example you want to fix, right?
> SELECT '%SystemDrive%\\Users\\John' _FUNC_ '\%SystemDrive\%\\\\Users%';

@dongjoon-hyun I would change this to don't mix 2 different things: how the SQL parser handles regular string literals and how `LIKE` handles patterns. Not obvious why those 2 escapes `/` and `\` lead to 2 different patterns:
```sql
spark-sql (default)> SELECT R'%SystemDrive%\Users\John' LIKE '\%SystemDrive\%\\\\Users%' ESCAPE R'\';
true
spark-sql (default)> SELECT R'%SystemDrive%/Users/John' LIKE '/%SystemDrive/%//Users%' ESCAPE R'/';
true
```

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:114]: LIKE and Java both handle `\` and only LIKE handles `/`.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:114]: > Use \ to escape special characters

So for `\a` that doesn't escape anything, `\` is treated as a normal char.

I think the problem here is the LIKE doc describes the behavior with the actual string value of the pattern. Can we reference the string literal doc here? And give examples about how to match `\abc`: `r'\\abc'` or `'\\\\abc'`.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:114]: I have updated the description of string literals in https://github.com/apache/spark/pull/43152, so, we can refer to the doc from here and other expression descriptions. Please, take a look at the PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala [hunk 2]
```diff
@@ -108,13 +108,15 @@ abstract class StringRegexExpression extends BinaryExpression
     Examples:
       > SELECT _FUNC_('Spark', '_park');
       true
+      > SELECT '\\abc' AS S, S _FUNC_ R'\\abc', S _FUNC_ '\\\\abc';
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala:111]: shall we improve the examples for other like functions as well?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala]: Added

**Review by srowen — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by srowen** [comment]
Just rerun tests, I guess, though doesn't look related. I'm surprised at the number of build issues we've had lately

**Comment by MaxGekk** [comment]
@srowen The failure of `ExpressionsSchemaSuite` is related to my changes. I have regenerated `sql-expression-schema.md` to fix it.

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @srowen @cloud-fan @beliefer @dongjoon-hyun for review.

**Pull request closed without merging**